### ISSUE TITLE
A parent shell app

### ISSUE THREAD

**Issue by prologic** [openmicroapps/meta#4]
Originally when I set out to build a set of micro-apps many of which can be found at my [Self-Hosted Golang apps by James Mills / prologic](https://gist.github.com/prologic/c165403fb6ac07c16dc9fd6eeb8efbd2) I had this idea that there would be (_which I didn't develop at the time_) an orch-arching parent / shell app that would encompass all the little micro-apps.

It would be responsible for:

- User Accounts
- User Preferences
- Switching between micro-apps

It would interact with the micro-apps by a defined protocol that would:

- Allow micro-apps "views" to be seamlessly embedded without switching away from the page/url/domain
- All data would be stored with the micro-app

**Comment by jtagcat** [comment]
Current apps (to be moved to openmicroapps):
 - [todo](https://github.com/prologic/todo)
 - [notes](https://github.com/prologic/notes)
 -  [wiki](https://github.com/prologic/wiki)
 -  [shorturl](https://github.com/prologic/shorturl)
 -  [pastebin](https://github.com/prologic/pastebin)

Planned: Some kind of numbers, data analysis, calculator, spreadsheet thing (but not a spreadsheet)

**Comment by jtagcat** [comment]
Probably gonna have it like domain.tld/todo, domain.tld/notes and so on, having it under the same domain simplifies cookies, caching and communication.

**Comment by prologic** [comment]
If we can flesh this out and build a prptotype we can also define what the protocols would look like. This would allow an ecosystem of other micro-apps to be built as we define the standard for how they should work.

**Comment by jtagcat** [comment]
Anything protocols are now at https://github.com/openmicroapps/unify along with discussion of the unify app itself.

**Comment by prologic** [comment]
👍
-- 

James Mills / prologic

E: prologic@shortcircuit.net.au
W: prologic.shortcircuit.net.au

**Comment by jtagcat** [comment]
@prologic please follow the rules https://github.com/openmicroapps/meta/blob/master/README.md
😆

**Comment by prologic** [comment]
Haha yeah okay! My laptop run out of battery so I’m being lazy and just
replying via phone
-- 

James Mills / prologic

E: prologic@shortcircuit.net.au
W: prologic.shortcircuit.net.au

**Comment by jtagcat** [comment]
This issue should be closed and the remains be split out in to other issues / readme.md updates.

**Comment by prologic** [comment]
Sounds good to me!  Have we captured all of the ideas here in separate
issues?
-- 

James Mills / prologic

E: prologic@shortcircuit.net.au
W: prologic.shortcircuit.net.au

**Comment by prologic** [comment]
We should probably think about and define the protocol of how the micro
apps are going to be embedded in the first place? Do we just pull in and
render their views?  How much client-side JavaScript do we want to use here
to achieve seamless integration?  Or do we try and do it all service side
at the Shell please and reroute requests to the micro apps?
-- 

James Mills / prologic

E: prologic@shortcircuit.net.au
W: prologic.shortcircuit.net.au

**Comment by jtagcat** [comment]
@prologic Don't exactly get what you mean here, but I think, we are trying to be (first of all) as fast as possible, and (secondly) as light as possible. Cache as much as we can, otherwise server-side (as long as it doesn't affect responsiveness too much).

**Comment by prologic** [comment]
Let me try again... There are two ways I think we can achieve a shell that encompases any micro-app:

Either we:

- Render each micro-app server-side using server<->server protocols between the shell and micro-app

or:

- We render the shell, some JS and expect client-side JS to "fetch the micro-app's code" and render/respond to user interactions.

Both have pros/cons -- The former would allow for a more graceful degradation design and users with low bandwidth. The later would be more seamless (_using react-style patterns, e;g: dom diff patching_) but would be considerably more complex to implement and heavier on resources.

**Comment by jtagcat** [comment]
Hmm, I really like the idea of the app first sending the requested, with todo the list's content, then at the same time client sends cookie, client chooses, if it's dark or light themed and app then gathers data, sends everything back to client.

..but at the same time the other option is attractive as well.

Since we are prototyping, could we first try both? 😏

**Comment by prologic** [comment]
We can definitely try both! The todo µApp should be our "test bed" :)

I can definitely do the first design I described (all done server-side) with minimal client-side JS mostly for aesthetics with graceful degradation.

Are you at all good with things like Ract, Redux, Preact or VirtualDOM? (https://www.sitepoint.com/react-alternatives-preact-virtualdom-deku/ is an interesting read on the subject)

**Comment by prologic** [comment]
I must point out though that going full React would definately go against the lightweight / minimal design originally set out :)
